### ISSUE TITLE
survival analysis / time-to-event analysis in numpyro

### ISSUE THREAD

**Issue by vanAmsterdam** [pyro-ppl/numpyro#1909]
in my research domain (healthcare) many outcomes are right-censored (aka survival analysis / time-to-event); the data for an individual patient are basically a tuple of (observed_time, event / censoring indicator).

What would be a good way to incorporate this into numpyro? There are several distributions already implemented in the library that are often used for (bayesian) survival analysis (e.g. exponential, weibull), but calculating the likelihood correctly for censored data would require some extra work.

are others interested in adding support for survival analysis to numpyro?

an example implemented in problow is here: https://probflow.readthedocs.io/en/stable/examples/time_to_event.html

**Comment by juanitorduz** [comment]
@vanAmsterdam There is an example where this is done by hand https://num.pyro.ai/en/latest/tutorials/censoring.html (motivated by the code in https://kylejcaron.github.io/posts/censored_demand/2024-02-06-censored-demand.html) . Would that be enough? Or are you looking into something like the PyMC wrapper https://www.pymc.io/projects/docs/en/latest/api/distributions/censored.html ?

**Comment by vanAmsterdam** [comment]
e.g. we might also think about it as an effect handler just like mask, but then a censored variant, and even potentially a truncated variant

so the interface could be something like

```python
t = jnp.array([.72, 1.23, .45])
censored = jnp.array([True, False, True])

with handlers.censored(censored):
   sample('obs_t', dist.Normal(), obs=t)

```

btw stan docs also have some nice background information on this topic: https://mc-stan.org/docs/stan-users-guide/truncation-censoring.html#censored-data

**Comment by juanitorduz** [comment]
> pymc approach to numpyro if we think this is a worthwhile
IMO, that would be great! This can enable users to use this more directly (also creating awareness about it). 

But I guess @fehiepsi and his team are the ones to decide whether this is something they want :)

**Comment by fehiepsi** [comment]
Yeah,we welcome the contribution.

**Comment by milesalanmoore** [comment]
@vanAmsterdam Any progress on this? I'm interested in participating if you're still working on this! I've been fitting (thresholding) survival models in Stan for my own work and agree this model class would be great in numpyro.

**Comment by vanAmsterdam** [comment]
update: I've now implemented Right / Left / Interval and Doubly censored distributions; you can check the files changed here: https://github.com/vanAmsterdam/numpyro/pull/1

a couple of tests are still failing and the documentation needs updating

**Comment by juanitorduz** [comment]
Hey! Thanks for the update! Regarding your points:

!. Do you have ideas on how to approach the tests?

I think (a personal opinion) we can go from simple to complex and work on iterations. We might want to merge the simplest case and then incrementally add more complexity (dimensions?) This is how we worked out the HSGP contrib module. It is also easier to review ;)

2. What to do with the notebooks.

I wrote the `censoring.ipynb` notebook, which I simply ported from my blog (https://juanitorduz.github.io/censoring/). Hence, you can simply replace it with your newer version and add the blog link if people want to go into more details.

**Comment by juanitorduz** [comment]
Regarding the notebook: the PyMC example was just an idea, so feel to choose or combine them :) No strong opinions here :) We can always iterate 🤜🤛
